drivers/thermal/thermal_procfs.c

```diff
@@ -0,0 +1,351 @@
+
+  level1 = kmm_zalloc(sizeof(struct procfs_dir_priv_s));
+  if (level1 == NULL)
+    {
+      *dir = NULL;
+      return -ENOMEM;
+    }
+
+  level1->level = 1;
+
+  nxmutex_lock(&g_thermal_procfs_lock);
+  level1->nentries = list_length(&g_thermal_procfs_list);
+  nxmutex_unlock(&g_thermal_procfs_lock);
+
+  *dir = (FAR struct fs_dirent_s *)level1;
+  return OK;
+}
+
+static int thermal_procfs_closedir(FAR struct fs_dirent_s *dir)
+{
+  kmm_free(dir);
+  return OK;
+}
+
+static int thermal_procfs_readdir(FAR struct fs_dirent_s *dir,
+                                  FAR struct dirent *entry)
+{
+  FAR struct procfs_dir_priv_s *level1;
+  FAR struct thermal_procfs_s *child;
+  int index = 0;
+
+  DEBUGASSERT(dir);
+  level1 = (FAR struct procfs_dir_priv_s *)dir;
+
+  if (level1->index >= level1->nentries)
+    {
+      /* We signal the end of the directory by returning the special
+       * error -ENOENT
+       */
+
+      return -ENOENT;
+    }
+
+  nxmutex_lock(&g_thermal_procfs_lock);
+
+  list_for_every_entry(&g_thermal_procfs_list, child,
+                       struct thermal_procfs_s, node)
+    {
+      if (index == level1->index)
+        {
+          entry->d_type = DTYPE_FILE;
+
+          strlcpy(entry->d_name, child->zdev->name, NAME_MAX);
+          level1->index++;
+          nxmutex_unlock(&g_thermal_procfs_lock);
+          return OK;
+        }
+
+      index++;
+    }
+
+  nxmutex_unlock(&g_thermal_procfs_lock);
+  return -ENOENT;
+}
+
+static int thermal_procfs_rewinddir(FAR struct fs_dirent_s *dir)
+{
+  FAR struct procfs_dir_priv_s *level1;
+
+  DEBUGASSERT(dir);
+  level1 = (FAR struct procfs_dir_priv_s *)dir;
+  level1->index = 0;
+  return OK;
+}
+
+static int thermal_procfs_stat(FAR const char *relpath,
+                               FAR struct stat *buf)
+{
+  FAR struct thermal_procfs_s *child;
+
+  memset(buf, 0, sizeof(struct stat));
+
+  if (strcmp(relpath, "thermal") == 0 || strcmp(relpath, "thermal/") == 0)
+    {
+      buf->st_mode = S_IFDIR | S_IROTH | S_IRGRP | S_IRUSR;
+      return OK;
+    }
+  else
+    {
+      relpath += strlen("thermal/");
+
+      nxmutex_lock(&g_thermal_procfs_lock);
+
+      list_for_every_entry(&g_thermal_procfs_list, child,
+                           struct thermal_procfs_s, node)
+        {
+          if (!strcmp(child->zdev->name, relpath))
+            {
+              buf->st_mode = S_IFREG | S_IROTH | S_IRGRP | S_IRUSR;
+              nxmutex_unlock(&g_thermal_procfs_lock);
+              return OK;
+            }
+        }
+
+      nxmutex_unlock(&g_thermal_procfs_lock);
+    }
+
+  return -ENOENT;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+int thermal_zone_procfs_register(FAR struct thermal_zone_device_s *zdev)
+{
+  FAR struct thermal_procfs_s *p;
+
+  p = kmm_zalloc(sizeof(struct thermal_procfs_s));
+  if (!p)
```

Comment:
```suggestion
  if (p == NULL)

Reply:
ACK, thanks, will upload it soon

Reply:
Done - fd06f178739718dccfa175e7ef9b3338a3ccbb53